### PR TITLE
Rewrite DirectoryWatcher

### PR DESCRIPTION
**Pull request by chihuahua**
Rewrite the `DirectoryWatcher` to support multiple `SummaryWriter`s writing
to the same run so long as the writers differ in events name prefix and
do not share common tags within events across writers.

This DirectoryWatcher is a work in progress.

See docs for the `DirectoryWatcher` constructor for details. See issue #1063.

TODO):
* Further flesh out the logic.
* This `DirectoryWatcher` works for a static mnist_with_summaries log
directory. It should be tried out more rigorously.
* Write tests.
* Periodically check for closed files that updated.

### CODE REVIEW

File: tensorboard/backend/event_processing/suffix_based_directory_watcher.py [hunk 1]
```diff
@@ -0,0 +1,272 @@
...
+  def _ParseSuffix(self, path):
+    """Parses the suffix.
+
+    Args:
+      path: The path to an events file.
+
+    Returns:
+      A string that is the suffix. Or none if the pattern does not match.
+    """
+    match = re.search('events\.out\.tfevents\.\d+\.(.*)$', path)
```
Comment (nfelt) [inline tensorboard/backend/event_processing/suffix_based_directory_watcher.py]: Add an `r` prefix so backslash escapes are valid.

File: tensorboard/backend/event_processing/suffix_based_directory_watcher.py [hunk 2]
```diff
@@ -0,0 +1,272 @@
...
+          if last_event:
+            at_least_1_event_read_for_current_paths = True
+            at_least_1_event_read = True
+            head.last_event_time = last_event.wall_time
+
+          # Close this file if either at least 1 events file (with the same
+          # suffix) comes after this events file or its last recorded event
+          # timestamp is old.
+          if (suffix_to_last_file_index[suffix] > i or (
+                head.last_event_time is not None and
```
Comment (nfelt) [inline tensorboard/backend/event_processing/suffix_based_directory_watcher.py]: Indent should only be +4 spaces here, not +6.  Ditto for line below.

File: tensorboard/backend/event_processing/suffix_based_directory_watcher.py (hunk 1 above)
Comment (chihuahua) [reply tensorboard/backend/event_processing/suffix_based_directory_watcher.py]: Done!

File: tensorboard/backend/event_processing/suffix_based_directory_watcher.py (hunk 2 above)
Comment (chihuahua) [reply tensorboard/backend/event_processing/suffix_based_directory_watcher.py]: Done!

**Comment by nfelt** [comment]
Yes, I think we should hold off on this for now (which is why I haven't merged it) even though I think it's likely that this functionality would be useful to have anyway, even if it isn't strictly necessary for the tf.contrib.summary migration.

**Comment by jart** [comment]
Even if we fix the file name fork problem, I agree it'd still be nice to have, possibly behind a flag. I'm not sure how much we want to rock the boat with event log scanning behavior. What TensorBoard did here has been rewritten by multiple internal services, Microsoft CNTK, DMLC Standalone TensorBoard, tensorboard-pytorch, and many other places.

**Comment by jart** [comment]
I'm closing this out, per meeting. We might not be able to make a change like this, at this time. We might have to do some type of plumbing in TensorFlow, so we don't have these v2 suffix files.

**Pull request closed without merging**